File: src/main/java/fortress/tfol/Signature.java
package fortress.tfol;

import fortress.data.PersistentSet;
import fortress.data.PersistentHashSet;
import fortress.data.PersistentInsertionOrderedHashSet;
import java.util.Optional;
import java.util.Set;
import fortress.util.Errors;

// Persistent and Immutable
// Internally consistent
public class Signature {
    private final PersistentSet<Type> types;
    private final PersistentSet<FuncDecl> functionDeclarations;
    private final PersistentSet<AnnotatedVar> constants;
    
    private Signature(PersistentSet<Type> types,
                      PersistentSet<FuncDecl> functionDeclarations,
                      PersistentSet<AnnotatedVar> constants) {
        this.types = types;
        this.functionDeclarations = functionDeclarations;
        this.constants = constants;
    }
    
    public static Signature empty() {
        // For testing consistency, use an insertion ordered set
        return new Signature(
            PersistentInsertionOrderedHashSet.empty().plus(Type.Bool), // Every signature contains Bool
            PersistentInsertionOrderedHashSet.empty(),
            PersistentInsertionOrderedHashSet.empty()
        );
    }
    
    public Signature withType(Type t) {
        assertTypeConsistent(t);
        return new Signature(types.plus(t), functionDeclarations, constants);
    }
    
    public Signature withTypes(Iterable<Type> types) {
        Signature sig = this;
        for(Type t : types) {
            sig = sig.withType(t);
        }
        return sig;
    }
    
    public Signature withTypes(Type... types) {
        Signature sig = this;
        for(Type t : types) {
            sig = sig.withType(t);
        }
        return sig;
    }
    
    public Signature withFunctionDeclaration(FuncDecl fdecl) {
        assertFuncDeclConsistent(fdecl);
        return new Signature(types, functionDeclarations.plus(fdecl), constants);
    }
    
    public Signature withFunctionDeclarations(Iterable<FuncDecl> fdecls) {
        Signature sig = this;
        for(FuncDecl f : fdecls) {
            sig = sig.withFunctionDeclaration(f);
        }
        return sig;
    }
    
    public Signature withFunctionDeclarations(FuncDecl... fdecls) {
        Signature sig = this;
        for(FuncDecl f : fdecls) {
            sig = sig.withFunctionDeclaration(f);
        }
        return sig;
    }
    
    public Signature withConstant(AnnotatedVar c) {
        assertConstConsistent(c);
        return new Signature(types, functionDeclarations, constants.plus(c));
    }
    
    public Signature withConstants(Iterable<AnnotatedVar> constants) {
        Signature sig = this;
        for(AnnotatedVar c : constants) {
            sig = sig.withConstant(c);
        }
        return sig;
    }
    
    public Signature withConstants(AnnotatedVar... constants) {
        Signature sig = this;
        for(AnnotatedVar c : constants) {
            sig = sig.withConstant(c);
        }
        return sig;
    }
    
    public Optional<AnnotatedVar> lookupConstant(Var v) {
        return constants.stream()
            .filter(c -> c.getVar().equals(v))
            .findAny();
    }
    
    public Optional<FuncDecl> lookupFunctionDeclaration(String functionName) {
        return functionDeclarations.stream()
            .filter(fdecl -> fdecl.getName().equals(functionName))
            .findAny();
    }
    
    public boolean hasType(String name) {
        return types.containsValue(Type.mkTypeConst(name));
    }
    
    public boolean hasType(Type type) {
        return types.containsValue(type);
    }
    
    /* Package private */ Set<AnnotatedVar> getConstants() {
        return constants;
    }
    
    /* Package private */ Set<FuncDecl> getFunctionDeclarations() {
        return functionDeclarations;
    } 
    
    /* Package private */ Set<Type> getTypes() {
        return types;
    } 
    
    private void assertTypeConsistent(Type t) {
        // Type must not share a name with any function
        Errors.precondition(! functionDeclarations.stream().anyMatch(
            (FuncDecl fdecl) -> fdecl.getName().equals(t.getName())
        ), "Name " + t.getName() + " shared by type and function");
        
        // Type must not share a name with any constant
        Errors.precondition(! constants.stream().anyMatch(
            (AnnotatedVar c) -> c.getName().equals(t.getName())
        ), "Name " + t.getName() + " shared by type and constant");
    }
    
    private void assertConstConsistent(AnnotatedVar c) {
        // Constant's type must be within the set of types
        Errors.precondition(types.containsValue(c.getType()),
            "Constant " + c.getName() + " of undeclared type " + c.getType().getName());
        
        // Constant's cannot share a name with a constant of a different type
        Errors.precondition(! constants.stream().anyMatch(
            (AnnotatedVar otherConst) -> otherConst.getName().equals(c.getName()) && !otherConst.equals(c)
        ), "Constant " + c.getName() + " declared with two different types");
        
        // Constant cannot share a name with any function 
        Errors.precondition(! functionDeclarations.stream().anyMatch(
            (FuncDecl fdecl) -> fdecl.getName().equals(c.getName())
        ), "Name " + c.getName() + " shared by constant and function");
    }
    
    private void assertFuncDeclConsistent(FuncDecl fdecl) {
        // Argument types must exist in type set
        Errors.precondition(fdecl.getArgTypes().stream().allMatch(types::containsValue),
            "Function " + fdecl.getName() + " has argument types that are undeclared");
            
        // Result type must exist in type set
        Errors.precondition(types.containsValue(fdecl.getResultType()),
            "Function " + fdecl.getName() + " has result type that is undeclared");
            
        // Function must not share name with a constant
        Errors.precondition(! constants.stream().anyMatch(
            (AnnotatedVar c) -> c.getName().equals(fdecl.getName())
        ), "Name " + fdecl.getName() +  " shared by function and constant");
        
        // Function must not share name with a type
        Errors.precondition(! types.stream().anyMatch(
            (Type type) -> type.getName().equals(fdecl.getName())
        ), "Name " + fdecl.getName() +  " shared by function and type");
        
        // Function must not share name with another function, unless it is the same function
        Errors.precondition(! functionDeclarations.stream().anyMatch(
            (FuncDecl otherFun) -> otherFun.getName().equals(fdecl.getName()) && ! otherFun.equals(fdecl)
        ), "Function " + fdecl.getName() + " declared with two different types");
    }
    
    @Override
    public boolean equals(Object other) {
        if(this == other) {
            return true;
        }
        if(null == other) {
            return false;
        }
        if(this.getClass() != other.getClass()) {
            return false;
        }
        Signature o = (Signature) other;
        return this.constants.equals(o.constants)
            && this.functionDeclarations.equals(o.functionDeclarations)
            && this.types.equals(o.types);
        
    }
    
    @Override
    public int hashCode() {
        int prime = 31;
        int result = 1;
        result = result * prime + constants.hashCode();
        result = result * prime + functionDeclarations.hashCode();
        result = result * prime + types.hashCode();
        return result;
    }
}

File: src/main/scala/fortress/tfol/Signature.scala
package fortress.tfol 

import fortress.data.InsertionOrderedSet
import fortress.util.Errors
import fortress.tfol.operations.TypeCheckResult
import scala.collection.JavaConverters._
import scala.annotation.varargs // So we can call Scala varargs methods from Java
import scala.collection.immutable.Seq // Use immutable seq by default

// Persistent and Immutable
// Internally consistent
// The constructor is private -- the only way to make signatures outside of this class
// is through the empty and withXYZ methods 
case class Signature private (
    sorts: Set[Sort],
    functionDeclarations: Set[FuncDecl],
    constants: Set[AnnotatedVar],
    enumConstants: Map[Sort, Seq[EnumValue]],
    extensions: Set[SignatureExtension]
) extends ExtensibleTypeChecking {
    
    // TODO need to check this type is not builtin
    def withSort(t: Sort): Signature = {
        assertSortConsistent(t)
        Signature(sorts + t, functionDeclarations, constants, enumConstants, extensions)
    }
    
    def withSorts(sorts: java.lang.Iterable[Sort]): Signature = {
        var sig: Signature = this
        sorts.forEach { t =>
            sig = sig.withSort(t)
        }
        sig
    }
    
    @varargs
    def withSorts(sorts: Sort*): Signature = withSorts(sorts.asJava)
    
    def withFunctionDeclaration(fdecl: FuncDecl): Signature = {
        assertFuncDeclConsistent(fdecl)
        Signature(sorts, functionDeclarations + fdecl, constants, enumConstants, extensions)
    }
    
    def withFunctionDeclarations(fdecls: java.lang.Iterable[FuncDecl]): Signature = {
        var sig: Signature = this
        fdecls.forEach { f =>
            sig = sig.withFunctionDeclaration(f)
        }
        sig
    }
    
    def withFunctionDeclarations(fdecls: Iterable[FuncDecl]): Signature = {
        var sig = this
        fdecls.foreach { f =>
            sig = sig.withFunctionDeclaration(f)
        }
        sig
    }
    
    @varargs
    def withFunctionDeclarations(fdecls: FuncDecl*): Signature = withFunctionDeclarations(fdecls.asJava)
    
    def withConstant(c: AnnotatedVar): Signature = {
        assertConstConsistent(c);
        Signature(sorts, functionDeclarations, constants + c, enumConstants, extensions)
    }
    
    def withConstants(constants: java.lang.Iterable[AnnotatedVar]): Signature = {
        var sig = this
        constants.forEach { c =>
            sig = sig.withConstant(c)
        }
        sig
    }
    
    def withConstants(constants: Iterable[AnnotatedVar]): Signature = {
        var sig = this
        constants.foreach { c =>
            sig = sig.withConstant(c)
        }
        sig;
    }
    
    @varargs
    def withConstants(constants: AnnotatedVar*): Signature = withConstants(constants.asJava)
    
    def withEnumSort(t: Sort, values: Seq[EnumValue]) = {
        // TODO more consistency checking
        Signature(sorts + t, functionDeclarations, constants, enumConstants + (t -> values), extensions)
    }
    
    def withEnumSort(t: Sort, values: java.util.List[EnumValue]) = {
        // TODO more consistency checking
        Signature(sorts + t, functionDeclarations, constants, enumConstants + (t -> values.asScala.toList), extensions)
    }
    
    // TypeChecking
    
    override def queryConstantBase(v: Var): Option[AnnotatedVar] = constants.find(_.variable == v)
    
    override def queryEnumBase(e: EnumValue): Option[Sort] = enumConstants.find {
        case (sort, enumConstants) => enumConstants contains e
    }.map { case (sort, _) => sort }
    
    override def queryFunctionBase(name: String, argSorts: Seq[Sort]): Option[FuncDecl] =
        functionDeclarations.find(fdecl => fdecl.name == name && fdecl.argSorts == argSorts)
    
    override def queryUninterpretedFunctionBase(name: String): Option[FuncDecl] = functionDeclarations.find(_.name == name)
    
    override def hasSortBase(sort: Sort): Boolean = extensions.exists(_.hasSort(sort)) || (sorts contains sort)
    
    override def hasSortWithNameBase(name: String): Boolean = extensions.exists(_.hasSortWithName(name)) || sorts.exists(_.name == name)
    
    override def hasFunctionWithNameBase(name: String): Boolean = (extensions.exists(_.hasFunctionWithName(name))) || functionDeclarations.exists(_.name == name)
    
    private[tfol]
    def getConstants: java.util.Set[AnnotatedVar] = constants.asJava
    
    private[tfol]
    def getFunctionDeclarations: java.util.Set[FuncDecl] = functionDeclarations.asJava
    
    private[tfol]
    def getSorts: java.util.Set[Sort] = sorts.asJava
    
    def withIntegers: Signature = Signature(sorts, functionDeclarations, constants, enumConstants, extensions + IntegerExtension)
    
    def withoutEnums = Signature(sorts, functionDeclarations, constants, Map.empty, extensions)
    
    private
    def assertSortConsistent(t: Sort): Unit = {
        // Sort must not share a name with any function
        Errors.precondition(! functionDeclarations.exists(
            (fdecl: FuncDecl) => fdecl.name == t.name
        ), "Name " + t.name + " shared by sort and function")
        
        // Sort must not share a name with any constant
        Errors.precondition(! constants.exists(
            (c: AnnotatedVar) => c.name == t.name
        ), "Name " + t.name + " shared by sort and constant")
    }
    
    private 
    def assertConstConsistent(c: AnnotatedVar): Unit = {
        // Constant's sort must be within the set of sorts
        Errors.precondition(sorts contains c.sort,
            "Constant " + c.toString + " of undeclared sort ")
        
        // Constant's cannot share a name with a constant of a different sorts
        Errors.precondition(! constants.exists(
            (otherConst: AnnotatedVar) => otherConst.name == c.name && otherConst != c
        ), "Constant " + c.name + " declared with two different sorts")
        
        // Constant cannot share a name with any function 
        Errors.precondition(! functionDeclarations.exists(
            (fdecl: FuncDecl) => fdecl.name == c.name
        ), "Name " + c.name + " shared by constant and function")
    }
    
    private
    def assertFuncDeclConsistent(fdecl: FuncDecl): Unit = {
        // Argument sorts must exist in sort set
        Errors.precondition(fdecl.argSorts.forall(t => sorts contains t),
            "Function " + fdecl.name + " has argument sorts that are undeclared")
            
        // Result sort must exist in sort set
        Errors.precondition(sorts.contains(fdecl.resultSort),
            "Function " + fdecl.name + " has result sort that is undeclared")
            
        // Function must not share name with a constant
        Errors.precondition(! constants.exists(
            (c: AnnotatedVar) => c.name == fdecl.name
        ), "Name " + fdecl.name +  " shared by function and constant")
        
        // Function must not share name with a sort
        Errors.precondition(! sorts.exists(
            (t: Sort) => t.name == fdecl.name
        ), "Name " + fdecl.name +  " shared by function and sort")
        
        // Function must not share name with another function, unless it is the same function
        Errors.precondition(! functionDeclarations.exists(
            (otherFun: FuncDecl) => otherFun.name == fdecl.name && otherFun != fdecl
        ), "Function " + fdecl.name + " declared with two different sorts")
    }
    
    override def toString: String = "Signature <<\n" + sorts.mkString("\n") + "\n" + functionDeclarations.mkString("\n") + "\n" + constants.mkString("\n") +
        "\nExtensions:\n" + extensions.mkString("\n") + ">>"
}

object Signature {
    def empty: Signature = 
        // For testing consistency for symmetry breaking, use an insertion ordered set
        Signature(InsertionOrderedSet.empty[Sort] + BoolSort, InsertionOrderedSet.empty, InsertionOrderedSet.empty, Map(), Set())
}
